### ISSUE TITLE
Error handler

### ISSUE THREAD

**Issue by trieloff** [adobe/hypermedia-pipeline#85]
From @kptdobe 
> The pipeline should have a global error handler, like `.error(error, status)` which allows you to centrally manage error cases. The pipeline is responsible to call the handler if any of the step returns an error.
> What do we do if the `pre.js` fires an error ? You must extend the pipeline and add a `when` right after to handle the case the pre function has a error case… ? That does not make sense.

**Comment by trieloff** [comment]
I would probably keep the `error` signature identical to all other signatures, so make it `error(context, action)`.

@kptdobe I assume that whatever `error` function you specify would unset the `context.error` property when it's done, correct?

If you force doing this manually, you could have multiple `error` functions, maybe even in combination with `when`, so that you can escalate fixes.

**Comment by kptdobe** [comment]
If there is an error that is not treated / caught in any of the action, this is a critical state and you must stop the pipeline execution and end up into the error handler.

I don't really know if error function would unset the context.error but at least if should properly log the error and make sure the proper response type is set correctly. Some use cases:

* With https://github.com/adobe/helix-cli/issues/302, it is clear that the developer forgot to catch an error, things should end up with a 500 and the appropriate log message.
* If `fetch` fails, there could be multiple reasons: md does not exist (404), wrong branch name (500 but user error), github might be unreachable (500 but unexpected error)... 

Some of those errors are the responsibility of the `fetch` method like non existing content (404), some falls into the category "user configuration error", some falls into "unexpected things happened" and ends up with a 500 that needs to be debugged.

The simplest thing for now is to route everything to a generic error function that logs the error and sets a 500 status if no status has been set before.

In dev, you only need the most accurate error message to debug.
In production, you may anyway only have 2 interesting cases: "not found" and "oops, something went wrong and there is nothing you can do about it" which are never cached.

**Comment by tripodsan** [comment]
### suggest:
- if a pipeline step throws an error, it sets the `payload.error`
- a step is ignored, if `payload.error` is set
- `pipeline.error()` is executed, if `payload.error` is set
- `pipeline.error()` usually sets the `response.status` if not already defined.
- `pipeline.error()` can clear the `payload.error` if needed (i.e. to implement a `try/catch` like semantics around some steps)

further, for the html pipeline:
- fetch should set the `response.status` to 404 **AND** the the `payload.error`
- we add `error(set-status)` as last step.

**Comment by tripodsan** [comment]
I'll address this together with #81 in #82

**Comment by trieloff** [comment]
@tripodsan I agree with your proposal, but not this part:

> a step is ignored, if `payload.error` is set

There are two ways to read this: "the next step is ignored" or "all steps are ignored as long as `payload.error` is set". In both cases there is too much magic for me, and it would be more transparent to simply add `.when(okay)` to the steps that should be skipped in case of errors.

**Comment by tripodsan** [comment]
> There are two ways to read this: "the next step is ignored" or "all steps are ignored as long as payload.error is set".

I don't think this is ambiguous: `if error then skip ...`

>  In both cases there is too much magic for me, and it would be more transparent to simply add .when(okay) to the steps that should be skipped in case of errors.

I disagree. Actually, I think error is error and the entire pipeline should just stop and throw an exception. The solution above is a tradeoff of this radical approach and allows 'some' error handling.

**Comment by trieloff** [comment]
As I wrote in my review in #82, I would prefer an approach that encourages error remediation, i.e. trying to fix things instead of just making the error prettier (i.e. setting correct status, or presenting error messages in HTML).

**Comment by tripodsan** [comment]
> I would prefer an approach that encourages error remediation

yes. this can be done with the current proposal. just add an `error()` step when you want to handle the error.

**Comment by tripodsan** [comment]
> or presenting error messages in HTML).

which is a very bad behaviour as it reveals too much information.
